File: lib/services/api_client.dart
// Thin http wrapper. Reads the JWT from shared_preferences, attaches it as a
// Bearer header, unwraps the backend's { success, message, data } envelope
// and turns non-2xx responses into ApiException with the backend's message.

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import 'api_config.dart';

class ApiException implements Exception {
  final int? status;
  final String message;
  ApiException(this.message, {this.status});
  @override
  String toString() => 'ApiException($status): $message';
}

class ApiClient {
  static const _tokenKey = 'auth_token';

  static Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenKey);
  }

  static Future<void> setToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, token);
  }

  static Future<void> clearToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove('auth_user');
  }

  static Uri _buildUri(String path, [Map<String, dynamic>? query]) {
    final base = Uri.parse(ApiConfig.apiBaseUrl);
    final full = path.startsWith('http')
        ? Uri.parse(path)
        : Uri.parse('${ApiConfig.apiBaseUrl}$path');
    if (query == null || query.isEmpty) return full;
    final qp = <String, String>{};
    query.forEach((k, v) {
      if (v != null) qp[k] = v.toString();
    });
    return full.replace(queryParameters: {
      ...full.queryParameters,
      ...qp,
    });
  }

  static Future<Map<String, String>> _headers({bool jsonBody = false}) async {
    final token = await getToken();
    return {
      if (jsonBody) 'Content-Type': 'application/json',
      'Accept': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  static dynamic _unwrap(http.Response res) {
    if (res.statusCode == 204) return null;
    dynamic body;
    try {
      body = res.body.isEmpty ? null : jsonDecode(res.body);
    } catch (_) {
      body = null;
    }
    if (res.statusCode >= 200 && res.statusCode < 300) {
      if (body is Map && body.containsKey('success')) {
        return body['data'];
      }
      return body;
    }
    final msg = (body is Map && body['message'] is String)
        ? body['message'] as String
        : 'HTTP ${res.statusCode}';
    throw ApiException(msg, status: res.statusCode);
  }

  static Future<dynamic> get(String path, {Map<String, dynamic>? query}) async {
    final res = await http.get(_buildUri(path, query), headers: await _headers());
    return _unwrap(res);
  }

  static Future<dynamic> post(String path, {Object? body}) async {
    final res = await http.post(
      _buildUri(path),
      headers: await _headers(jsonBody: body != null),
      body: body == null ? null : jsonEncode(body),
    );
    return _unwrap(res);
  }

  static Future<dynamic> put(String path, {Object? body}) async {
    final res = await http.put(
      _buildUri(path),
      headers: await _headers(jsonBody: body != null),
      body: body == null ? null : jsonEncode(body),
    );
    return _unwrap(res);
  }

  static Future<dynamic> patch(String path, {Object? body}) async {
    final res = await http.patch(
      _buildUri(path),
      headers: await _headers(jsonBody: body != null),
      body: body == null ? null : jsonEncode(body),
    );
    return _unwrap(res);
  }

  static Future<dynamic> delete(String path) async {
    final res = await http.delete(
      _buildUri(path),
      headers: await _headers(),
    );
    return _unwrap(res);
  }
}

File: react-frontend/src/services/apiClient.js
// Thin fetch wrapper. Domain services call this when USE_MOCK is false.
// Unwraps the backend's { success, message, data } envelope and surfaces
// backend error messages on non-2xx responses.

import { API_BASE_URL } from './config';

const buildUrl = (path, query) => {
  const url = new URL(
    path.startsWith('http') ? path : `${API_BASE_URL}${path}`,
  );
  if (query) {
    for (const [k, v] of Object.entries(query)) {
      if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
    }
  }
  return url.toString();
};

const getAuthToken = () =>
  typeof window === 'undefined' ? null : localStorage.getItem('auth_token');

// Returns the raw fetch Response. Most callers want apiRequest(), which
// unwraps the JSON envelope; export download endpoints use this directly.
export const apiFetch = async (path, options = {}) => {
  const { query, headers, ...init } = options;
  const token = getAuthToken();
  return fetch(buildUrl(path, query), {
    ...init,
    headers: {
      ...(init.body ? { 'Content-Type': 'application/json' } : {}),
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...(headers ?? {}),
    },
  });
};

export const apiRequest = async (path, options = {}) => {
  const res = await apiFetch(path, options);

  if (res.status === 204) return undefined;

  const ct = res.headers.get('content-type') ?? '';
  const isJson = ct.includes('application/json');
  const body = isJson ? await res.json().catch(() => null) : null;

  if (!res.ok) {
    const msg = body?.message || res.statusText || `HTTP ${res.status}`;
    throw new Error(msg);
  }

  // Backend wraps responses as { success, message, data }. Unwrap when present;
  // otherwise return the body as-is (some endpoints may return raw JSON).
  if (body && typeof body === 'object' && 'success' in body) {
    return body.data;
  }
  return body;
};

// Simulate latency for mock responses so the UI shows loading states realistically.
export const mockDelay = (ms = 250) =>
  new Promise((r) => setTimeout(r, ms));
